doc/website/blog/2022-04-07-how-twitter-can-implement.md

```diff
@@ -0,0 +1,127 @@
+
+| id      | tweet_id | timestamp | event | content |
+| ----------- | ----------- | ----------- | ----------- | ----------- |
+| 1      | 1       | 2022-04-06T12:30:00+00:00       | CREATED       | hello world!       |
+| 2      | 2       | 2022-04-06T13:45:34+00:00       | UPDATED       | hello Twitter!       |
+
+With a table similar to the one above, we can record changes to the original tweet "1" and if requested, we can show that it was originally tweeted at 12:30:00 with the content "hello world!" and was modified at 13:45:34 to "hello Twitter!".  
+
+To implement this, we will have to change every `UPDATE` statement for "tweets" to include an `INSERT` to "tweets_history". For correctness, we will need to wrap both statements in a transaction to avoid corrupting the history. in case the first statement succeeds but the subsequent one fails. We'd also need to make sure every `INSERT` to "tweets" is coupled with an `INSERT` to "tweets_history"
+
+```diff
+# INSERT is logged as "CREATE" history event
+- INSERT INTO tweets (`content`) VALUES ('Hello World!');
++BEGIN;
++INSERT INTO tweets (`content`) VALUES ('Hello World!');
++INSERT INTO tweets_history (`content`, `timestamp`, `record_id`, `event`)
++VALUES ('Hello World!', NOW(), 1, 'CREATE');
++COMMIT;
+
+# UPDATE is logged as "UPDATE" history event
+- UPDATE tweets SET `content` = 'Hello World!' WHERE id = 1;
++BEGIN;
++UPDATE tweets SET `content` = 'Hello World!' WHERE id = 1;
++INSERT INTO tweets_history (`content`, `timestamp`, `record_id`, `event`)
++VALUES ('Hello World!', NOW(), 1, 'UPDATE');
++COMMIT;
+```
+
+This method is nice but you'd have to create another table for different entities ("comment_history", "settings_history"). To prevent that, Enthistory creates a single "history" and a single "changes" table and records all the tracked fields there. It also supports many type of fields without needing to add more columns.
+
+### Pre release
+Enthistory is still in early design stages and is being internally tested. Therefore, we haven't released it to open-source yet - but we plan to do so very soon.
+If you want to play with a pre-release version of Enthistory, I wrote a simple React application with GraphQL+Enthistory to demonstrate how a tweet edit could look like. You can check it out [here](https://github.com/hedwigz/edit-twitter-example-app). please feel free to share your feedback.
+
+### Wrapping up
+We saw how Ent's modular extension system lets you streamline advanced features as if they were just a package install away. Developing your own extension [is fun, easy and educating](https://entgo.io/blog/2021/12/09/contributing-my-first-feature-to-ent-grpc-plugin)! I invite you to try it yourself!
+In the future, Enthistory will be used to track changes to Edges (aka Foreign keyed tables), generate automatic GraphQL schemas and mapping functions, and provide more methods for its underlying implementation.
```

Comment:
How is Enthistory related to graphQL?

Reply:
Do you mean generate GraphQL schemas to query the history? 

I think you can say an integration with the GraphQL and OpenAPI extensions to automatically generate endpoints for serving entity history to users